### ISSUE TITLE
Not able to set Title or Author via YAML Frontmatter

### ISSUE THREAD

**Issue by xge** [Zettlr/Zettlr#1401]
## Description
I am not able to set title or author via the YAML Frontmatter block. These get overwritten by the defaults set in the Project Settings dialog.
I verified that frontmatter is working by setting the `reference-section-title` variable.
I also tried setting author and title to an empty string, but the dialog won't let me save that.

## Reproducing
Add `title` or `author` variables in YAML Frontmatter and export the project via `Export Project`.
Exported PDF has Author and Title from Project Settings dialog, not from YAML Frontmatter.

## Expected behaviour

I want to be able to generate a title page with the author (+affiliations) and title defined in the YAML Frontmatter.

## Platform
 - OS and version: Windows 10
 - Zettlr Version: 1.7.5 and 1.8.0-beta.1

**Comment by nathanlesage** [comment]
Ah, it might very well be that the project settings override frontmatter vars, but that is actually desired behaviour, I suppose, as the primary aim for a project is to combine multiple files, hence certain frontmatter variables such as titles should be set globally, not locally. Do you happen to use projects for simply using different templates?

**Comment by xge** [comment]
I use projects to split long documents into multiple files.

**Comment by nathanlesage** [comment]
Yes, you are right… but I don't know how to implement this cleverly from the top of my head

**Comment by the-solipsist** [comment]
I don't know how exactly projects are implemented in Zettlr, but one potential way could be for projects settings to be stored in a Pandoc "defaults" file.

As far as I can see, the project settings cover:

- Project title
- PDF author
- PDF keywords
- TeX template
- TOC
- TOC depth level
- Generate title page (?)
- CSL
- Export format
- Margins
- Pagesize
- Page numbering
- Fonts
- Line spacing.

As far as I can know, except for "Generate title page", all the rest can be set as metadata or variables in a Pandoc "defaults" file.  So, a simple way to do this would be for Zettlr to create a YAML file ("projectname-defaults.yaml") based on all of these settings in the project directory.  A user who needs greater control (such as multiple authors, etc.) can directly edit the YAML file and add more fields and values.  Most users won't need to.  Zettlr can call this using `-d projectname-defaults.yaml`.

The other alternative (more in line with @xge's needs) is to create a "projectname-metadata.yaml" file, and call that as a `--metadata-file projectname-metadata.yaml`.  This option would allow users who wish to override the project settings to do so with in-file YAML.

I don't know if this is possible, but if it is, then this would be one way of addressing the needs of normal users and power users at the same time without adding greater complexity and clutter to the UI.

**Comment by Kangie** [comment]
To jump in here, Zettlr should respect YAML frontmatter when setting default fields. I'd even argue that the content should override whatever has been set at the project level - if it's in the document it should take priority. 

This is consistent with the way that pandoc handles things (from memory), as variables in content are parsed after those passed in the pandoc command. 

The 'create title page' could be an if statement, if nothing is passed the variable will be empty and not acted on - I have a few templates that I use in this way with pandoc. Gotta run, but honestly the best solution (that I can see) is moving the titlepage content into the template and letting Zettlr pass variables to pandoc during the command, and having pandoc read the frontmatter. 

Some testing will be required to see how this works. I'll see what I can do about validating which has precedence - command line or frontmatter.

**Comment by nathanlesage** [comment]
> Gotta run, but honestly the best solution (that I can see) is moving the titlepage content into the template and letting Zettlr pass variables to pandoc during the command, and having pandoc read the frontmatter.

This is on my bucket list either way, but this will come with a bigger amount of features, among them a built-in binary of Pandoc shipped with the Zettlr binary. This will close a whole bunch of issues but requires me not to monkey-patch but rather create a whole new module that is able to deal with the CLI flags to the Pandoc binary. I've been meaning to do this either way for some time, and it's very far up on the priority list.

Please do, in the meantime, attempt to collect as much info as possible so I can take all of those precedence-settings of Pandoc into account and build something lasting!

**Comment by the-solipsist** [comment]
What do you mean by "precedence-settings"?  If by that you mean the order of precedence of various variables and metadata options as set in different places, then there is no real documentation on this so far, so it can only be determined by either reading Pandoc's code or through [trial and error, as I've tried](https://groups.google.com/g/pandoc-discuss/c/nBmqXmmqULg/m/eJNYwV0XBQAJ).  I welcome others to contribute to that discussion.

**Comment by nathanlesage** [comment]
@tarleb Do you have an answer to @the-solipsist's question? That would be awesome <3

**Comment by tarleb** [comment]
In reverse order of precedence, i.e. later metadata setting override earlier ones:

- defaults files;
- in-document YAML block(s); later blocks take precedence over earlier blocks;
- command line settings (specifying the same key twice creates a list that contains both values).

Variables always have precedence over metadata fields when rendering.

This is from the top of my head, so it might be incomplete or partially wrong. I agree that this needs better documentation.

**Comment by Kangie** [comment]
My recent experience indicates that (at least for toc) YAML overrides the pandoc command. 

I actually need to document this behaviour for work, so stay tuned!

**Comment by nathanlesage** [comment]
> I actually need to document this behaviour for work, so stay tuned!

Sounds great! This would be incredibly helpful for our own Pandoc integration (+ if it makes it into the Pandoc docs for everyone else) :)

**Comment by tarleb** [comment]
I had a look at the sources, so let me try again. Again, in decreasing order of precedence

- metadata from defaults file (via `--defaults`)
- metadata from dedicated metadata file (via `--metadata-file`)
- metadata from command line (via `--metadata` or `-M`; repeating the same key creates an array)
- metadata from the document (e.g., YAML blocks; keys in later blocks take precedence)

@Kangie It would be great if you could get permission to feed this back into the pandoc docs.

**Comment by Kangie** [comment]
I'll see what I can do. Can't see it being commercially sensitive! Worst case I'll have to recreate my results on the taking some extra time. 

What you've just outlined broadly corresponds with my experience, particularly where in-document YAML blocks take precedence over the command line.

**Comment by Kangie** [comment]
OK, So I'm still working (slowly) on this - got pulled aside to implement a documentation pipeline, so it went on the backburner for a bit. I've hit the point though where I finally have both time and a pressing need to properly dig into it. Hopefully I should have something a bit better next week, or in early Jan, depending.

What I've seen so far broadly aligns with what @tarleb said, but I think that there's some variables/metadata (e.g. `toc` & `tocdepth`) where the YAML frontmatter seems to take precedence over the command line. I'll be digging into the code and running the current version of pandoc through some tests, so if there's something specific that you're curious about let me know. I need to decide if my documentation build system should extract frontmatter into a dedicated metadata file internally to give the user more control!

**Comment by the-solipsist** [comment]
> What I've seen so far broadly aligns with what @tarleb said, but I think that there's some variables/metadata (e.g. `toc` & `tocdepth`) where the YAML frontmatter seems to take precedence over the command line. I'll be digging into the code and running the current version of pandoc through some tests, so if there's something specific that you're curious about let me know.

I think I ran into a related issue earlier, and brought it up in [this discussion](https://groups.google.com/g/pandoc-discuss/c/tqMp0UKPpZQ/m/lLn3HkgkBwAJ) on the pandoc mailing list.

I was trying to figure out the difference between metadata field keys, (commandline) options, and variables, as set in defaults files.  I noted:

> I noticed something strange with "toc"/"table-of-contents":
> If I set "table-of-contents: true" as a metadata key/value or variable,
> nothing happens. If I set "toc: true" as a metadata field key/value or
> variable, then "toc-title" variable gets printed in the HTML output (so
> "Contents" shows up) but no table of contents appears in HTML, but a table
> of contents does appear in PDF! If I set "toc"/"table-of-contents" as an
> option, then the TOC shows up in both HTML and PDF. I feel this may be a
> flaw in the HTML5 template. 

@jgm replied:

> Best to regard the toc variable as for internal use only. In
some formats it does something, in others not, and what it does
can be different, due to differences in the formats. In LaTeX,
all we need for a table of contents is a single command in the
preamble; this can be set with a template so `toc: true`
suffices. In HTML, the `toc` variable activates a toc section in the
template, but the contents of this section must be generated
by the writer, which only happens if you selected `--toc` on
the command line. That is the documented way to do it.

**Comment by jkeiren** [comment]
As far as I can tell, the title is not even properly exported when exporting a document with YAML frontmatter through the "Share the current file as..." button (or File -> Export). 

If I take a file with the following content:

```
---
title: "Zettlr export title test"
...

Test document.
```

 Both export to HTML and PDF results in a document with no title in the document itself. Note, however that the `<title></title>` tag in the HTML export is set correctly as "Zettlr export title test". 

So, as far as I can tell, even without a project, the title (and author) in the PDF are not set according to the YAML frontmatter.

**Comment by nathanlesage** [comment]
Holy, there has been quite the discussion here while I attempted to ignore the issue list (and failed)!

So what I've taken away is the following:

- Zettlr should, in the future, write everything that's been set using dialogs into a temporary defaults-file that gets passed to Pandoc instead of command line arguments for the most part.
- We definitely need to overwork the default templates Zettlr ships
- We need Pandoc bundled quickly :D

But I now realised the problems, I think the easiest mitigation for now is to simply pass the Pandoc default template as the "custom template" wherever you want to export to PDF, as that one will make use of all the Pandoc behaviour you've just documented.

Furthermore, I accidentally introduced some errors into the Pandoc command, and the whole mess with the 2.10 --> 2.11-switch is getting over my head.

I'm going to label this as "wont fix" and instead focus on the really necessary bundling of Pandoc instead. So please use the Pandoc default template for now, and direct any comments to #1552 so that I can take note of them.
